base/compiler/typelimits.jl

```diff
@@ -559,17 +559,24 @@ function tmeet(@nospecialize(v), @nospecialize(t))
             return Bottom
         end
         @assert widev <: Tuple
-        new_fields = Vector{Any}(undef, length(v.fields))
-        for i = 1:length(new_fields)
-            if isa(v.fields[i], Core.TypeofVararg)
-                new_fields[i] = v.fields[i]
-            else
-                new_fields[i] = tmeet(v.fields[i], widenconst(getfield_tfunc(t, Const(i))))
-                if new_fields[i] === Bottom
-                    return Bottom
-                end
+        if isa(ti, DataType) && ti.name === Tuple.name
+            num_fields = length(ti.parameters)
+            isva = isvatuple(ti)
+        else
+            nfields_ti = nfields_tfunc(ti)
+            isva = !isa(nfields_ti, Const)
+            num_fields = isva ? length(v.fields) : (nfields_ti::Const).val
+        end
+        new_fields = Vector{Any}(undef, num_fields)
+        for i = 1:num_fields
+            new_fields[i] = tmeet(getfield_tfunc(v, Const(i)), widenconst(getfield_tfunc(ti, Const(i))))
```

Comment:
`t` is actually expected to be a slightly stronger constraint here. We already know that the LHS is already <: widev (by original construction), so there's nothing to be gained by computing the `tmeet` with it. There's a (small) chance that typeintersect was approximate however, and thus wider than `t`.

Reply:
By similar logic, the final return operation of this function `typeintersect(widenconst(v), t)` can also be strengthen:
```
...
elseif isa(v, Type)
    return typeintersect(v, t)
elseif !has_free_typevars(t) && typeintersect(widenconst(v), t) === Union{}
    return Bottom
end
return v # this leaf type can't be improved by intersection with type `t`
```
(which also subsumes the conditionals for Const and Conditional)

Reply:
> `t` is actually expected to be a slightly stronger constraint here. 

Um, yes, I had been pondering `t` vs. `ti` for a moment here, basically came to the same conclusion, and then chose `ti` for reasons I can't remember. I'll change this to `t` (or report back in case I remember what made we pick `ti` instead).

Reply:
Ah, yes, I've even added a test for this: if `v.typ == Tuple{Int,Vararg{Number}}` and `t = Union{Tuple{Int, Int, Int}, Tuple{Vararg{Float64}}}`, then `ti == Tuple{Int, Int, Int}`. And `tmeet(Number, getfield_tfunc(ti, Const(2)) == Int` while `tmeet(Number, getfield_tfunc(t, Const(2)) == Union{Int,Float64}`. (And this doesn't depend on `Vararg`, that's just what the test does.)
Admittedly, a `Union` on the RHS of a type-assert might not be to worry about too much, so maybe still better to use `t`, live with the imprecision in this case but be safe against approximation in intersection? Not sure.

Reply:
Regarding the second comment (about the final return), I took that as inspiration for a slightly different restructuring of the code. Please take a look whether you like. That also made m wonder about the handling of `t` with free typevars here, see the respective comment.

Reply:
`getfield_tfunc(v, Const(i))` seems so much more costly than `v.fields[i]`, though perhaps it isn't significant overall. How about:


```suggestion
new_fields[i] = tmeet(unwrapva(v.fields[min(i, end)]), widenconst(getfield_tfunc(ti, Const(i))))
```

Reply:
Included in rebase.